File: aoc2025-go/days/day07/day07.go
package day07

import (
	"fmt"
	"time"

	"aoc2025-go/common"
	"aoc2025-go/parse"
)

type Position struct {
	row, col int
}

func getStart(board []string) Position {
	for row := range board {
		for col, c := range board[row] {
			if c == 'S' {
				return Position{row, col}
			}
		}
	}
	panic("no start")
}

func inBounds(board []string, row, col int) bool {
	return row >= 0 && row < len(board) && col >= 0 && col < len(board[row])
}

func countSplits(board []string, pos Position, visited map[Position]struct{}) int {
	splits := 0
	for {
		if _, seen := visited[pos]; seen {
			return splits
		}
		visited[pos] = struct{}{}

		nextRow := pos.row + 1
		if !inBounds(board, nextRow, pos.col) {
			return splits
		}

		if board[nextRow][pos.col] == '^' {
			splits++
			if inBounds(board, nextRow, pos.col-1) {
				splits += countSplits(board, Position{nextRow, pos.col - 1}, visited)
			}
			if inBounds(board, nextRow, pos.col+1) {
				splits += countSplits(board, Position{nextRow, pos.col + 1}, visited)
			}
			return splits
		}
		pos.row = nextRow
	}
}

func countPaths(board []string, pos Position, memo map[Position]int) int {
	var straight []Position
	for {
		if val, exists := memo[pos]; exists {
			for _, p := range straight {
				memo[p] = val
			}
			return val
		}

		nextRow := pos.row + 1
		if !inBounds(board, nextRow, pos.col) {
			for _, p := range straight {
				memo[p] = 1
			}
			memo[pos] = 1
			return 1
		}

		if board[nextRow][pos.col] == '^' {
			result := 0
			if inBounds(board, nextRow, pos.col-1) {
				result += countPaths(board, Position{nextRow, pos.col - 1}, memo)
			}
			if inBounds(board, nextRow, pos.col+1) {
				result += countPaths(board, Position{nextRow, pos.col + 1}, memo)
			}
			memo[pos] = result
			for _, p := range straight {
				memo[p] = result
			}
			return result
		}

		straight = append(straight, pos)
		pos.row = nextRow
	}
}

func Part1(board []string) string {
	visited := make(map[Position]struct{})
	splits := countSplits(board, getStart(board), visited)
	return fmt.Sprintf("%d", splits)
}

func Part2(board []string) string {
	memo := make(map[Position]int)
	count := countPaths(board, getStart(board), memo)
	return fmt.Sprintf("%d", count)
}

func Run() {
	input := parse.Lines(7, false)

	t1s := time.Now()
	p1 := Part1(input)
	t1 := time.Since(t1s)

	t2s := time.Now()
	p2 := Part2(input)
	t2 := time.Since(t2s)

	common.RunDay(7, p1, p2, t1, t2)
}

File: aoc2025-zig/src/days/day07.zig
const std = @import("std");
const common = @import("../common.zig");
const parse = @import("../parse.zig");

const Position = struct {
    row: usize,
    col: usize,
};

fn getStart(board: [][]const u8) Position {
    for (board, 0..) |line, row| {
        for (line, 0..) |c, col| {
            if (c == 'S') return .{ .row = row, .col = col };
        }
    }
    unreachable;
}

fn inBounds(board: [][]const u8, row: isize, col: isize) bool {
    if (row < 0 or col < 0) return false;
    const r: usize = @intCast(row);
    if (r >= board.len) return false;
    const c: usize = @intCast(col);
    return c < board[r].len;
}

fn countSplits(
    board: [][]const u8,
    start: Position,
    visited: *std.AutoHashMap(Position, void),
    allocator: std.mem.Allocator,
) !usize {
    var pos = start;
    var splits: usize = 0;
    while (true) {
        const entry = try visited.getOrPut(pos);
        if (entry.found_existing) return splits;

        const next_row: isize = @as(isize, @intCast(pos.row)) + 1;
        if (!inBounds(board, next_row, @intCast(pos.col))) return splits;

        const nr: usize = @intCast(next_row);
        if (board[nr][pos.col] == '^') {
            splits += 1;
            const lc: isize = @as(isize, @intCast(pos.col)) - 1;
            const rc: isize = @as(isize, @intCast(pos.col)) + 1;
            if (inBounds(board, next_row, lc)) {
                splits += try countSplits(board, .{ .row = nr, .col = @intCast(lc) }, visited, allocator);
            }
            if (inBounds(board, next_row, rc)) {
                splits += try countSplits(board, .{ .row = nr, .col = @intCast(rc) }, visited, allocator);
            }
            return splits;
        }
        pos.row = nr;
    }
}

fn countPaths(
    board: [][]const u8,
    start: Position,
    memo: *std.AutoHashMap(Position, usize),
    allocator: std.mem.Allocator,
) !usize {
    var straight: std.ArrayList(Position) = .empty;
    defer straight.deinit(allocator);

    var pos = start;
    while (true) {
        if (memo.get(pos)) |val| {
            for (straight.items) |p| try memo.put(p, val);
            return val;
        }

        const next_row: isize = @as(isize, @intCast(pos.row)) + 1;
        if (!inBounds(board, next_row, @intCast(pos.col))) {
            for (straight.items) |p| try memo.put(p, 1);
            try memo.put(pos, 1);
            return 1;
        }

        const nr: usize = @intCast(next_row);
        if (board[nr][pos.col] == '^') {
            var result: usize = 0;
            const lc: isize = @as(isize, @intCast(pos.col)) - 1;
            const rc: isize = @as(isize, @intCast(pos.col)) + 1;
            if (inBounds(board, next_row, lc)) {
                result += try countPaths(board, .{ .row = nr, .col = @intCast(lc) }, memo, allocator);
            }
            if (inBounds(board, next_row, rc)) {
                result += try countPaths(board, .{ .row = nr, .col = @intCast(rc) }, memo, allocator);
            }
            try memo.put(pos, result);
            for (straight.items) |p| try memo.put(p, result);
            return result;
        }

        try straight.append(allocator, pos);
        pos.row = nr;
    }
}

pub fn part1(input: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    var visited = std.AutoHashMap(Position, void).init(allocator);
    defer visited.deinit();
    const splits = try countSplits(input, getStart(input), &visited, allocator);
    return std.fmt.allocPrint(allocator, "{d}", .{splits});
}

pub fn part2(input: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    var memo = std.AutoHashMap(Position, usize).init(allocator);
    defer memo.deinit();
    const count = try countPaths(input, getStart(input), &memo, allocator);
    return std.fmt.allocPrint(allocator, "{d}", .{count});
}

pub fn run(io: std.Io, allocator: std.mem.Allocator) !void {
    const input = try parse.lines(io, allocator, 7, false);

    const t1s = std.Io.Timestamp.now(io, .awake);
    const p1 = try part1(input, allocator);
    const t1e = std.Io.Timestamp.now(io, .awake);

    const t2s = std.Io.Timestamp.now(io, .awake);
    const p2 = try part2(input, allocator);
    const t2e = std.Io.Timestamp.now(io, .awake);

    common.runDay(io, 7, p1, p2, t1s, t1e, t2s, t2e);
}
